modules/combined/test/tests/mechanical_contact_constraint/blocks_2d/frictionless_kinematic_gap_offsets.i

```diff
@@ -0,0 +1,195 @@
+# this test is the same as frictionless_kinematic test but designed to test the gap offset capability
+# gap offsets with value of 0.01 were introduced to both master and slave sides in the initial mesh
+# these values were accounted using the gap offset capability to produce the same result as if no gap offsets were introduced
+[Mesh]
+  file = blocks_2d_gap_offset.e
+[]
+
+[GlobalParams]
+  displacements = 'disp_x disp_y'
+[]
+
+[AuxVariables]
+  [./penetration]
+  [../]
+  [./inc_slip_x]
+  [../]
+  [./inc_slip_y]
+  [../]
+  [./accum_slip_x]
+  [../]
+  [./accum_slip_y]
+  [../]
+  [./master_gap_offset]
+  [../]
+  [./slave_gap_offset]
+  [../]
+  [./mapped_master_gap_offset]
+  [../]
+[]
+
+[Functions]
+  [./vertical_movement]
+    type = ParsedFunction
+    value = -t
+  [../]
+[]
+
+[Modules/TensorMechanics/Master]
+  [./all]
+    add_variables = true
+    strain = FINITE
+  [../]
+[]
+
+[AuxKernels]
+  [./zeroslip_x]
+    type = ConstantAux
+    variable = inc_slip_x
+    boundary = 3
+    execute_on = timestep_begin
+    value = 0.0
+  [../]
+  [./zeroslip_y]
+    type = ConstantAux
+    variable = inc_slip_y
+    boundary = 3
+    execute_on = timestep_begin
+    value = 0.0
+  [../]
+  [./accum_slip_x]
+    type = AccumulateAux
+    variable = accum_slip_x
+    accumulate_from_variable = inc_slip_x
+    execute_on = timestep_end
+  [../]
+  [./accum_slip_y]
+    type = AccumulateAux
+    variable = accum_slip_y
+    accumulate_from_variable = inc_slip_y
+    execute_on = timestep_end
+  [../]
+  [./penetration]
+    type = PenetrationAux
+    variable = penetration
+    boundary = 3
+    paired_boundary = 2
+    slave_gap_offset = slave_gap_offset
```

Comment:
Overall, I think this PR is good to go.  The one nagging concern I have is that it seems pretty error-prone that the user will have to specify these offsets in two places. I guess this only comes up if they are separately asking for the penetration for output. I wonder if there's some way we could set these offsets somehow in the data structure that stores the penetration info and have PenetrationAux look that up so we don't have to have that be a parameter.

Reply:
I think this has something to do with how the PenetrationAux interacts with the contact enforcement classes. For the thermal contact, I think we only need to specify the gap_offset in the PenetrationAux, and the thermal contact classes will use the penetration value returned from PenetrationAux to enforce the thermal contact. However, in mechanical contact, the contact enforcement classes don't use the penetration value returned from the PenetrationAux but do its own calculation, and this is why we also need to provide gap_offset for the mechanical contact enforcement classes. You are right that for mechanical contact, we will need to provide the gap_offset for both the PenetrationAux and Contact action objects. We need to keep the thermal contact in mind while trying to avoid providing twice the gap_offset variable for the mechanical contact.

Reply:
It seems to me like one of the big problems here is that the `Contact` action sets up `PenetrationAux`, but `ThermalContact` doesn't.

Would it make sense to have `PenetrationAux` always provide the uncorrected distance between the faces, and have thermal contact do the adjustment internally?  This would make thermal and mechanical contact more consistent with each other.

Reply:
Surely we can make changes to `ThermalContact` to be consistent with `Contact` action. But I don't think this will be a clean solution. There will be at least two potential issues: 1) whenever we want to inquire any info using `PenetrationAux`, the results will be incorrect for cases with gap offsets; 2) this will introduce another related issues that if we want to build new objects based on gap distance, we will need to reimplement the actual gap distance in a way similar to that in the `Contact` action rather than directly get the value from `PenetrationAux`.

Reply:
I agree that it would be better for `PenetrationAux` to always report the penetration with the offsets taken into account. If we could have that taken into account behind the scenes in the GeometricSearch, which is set up by the Contact and ThermalContact actions, maybe we wouldn't have this problem.